### DISCUSSION TITLE
accessing array elements in function

### DISCUSSION THREAD

**Discussion by lg32810** [ezylang/EvalEx#415] [Q&A]
Hi everyone. I am trying to create a function that compute the sum of structured elements in a list ( list of hashmap ), I would like to express that like : SUMALL( list[item.amount] )  SUMALL( list[*.amount] ) but there is no  way apparently. Am I wrong ? As a first step, I added (successfully) a function to get the size à a list : 
	@FunctionParameter(name = "list")
	static public class CollectionSizeFunction extends AbstractFunction {
		@Override
		public EvaluationValue evaluate(Expression expression, Token functionToken,
				EvaluationValue... parameterValues) {
			return expression.convertValue(parameterValues[0].getArrayValue().size());
		}
	}
Thanks for your help,
Phil

**Comment by uklimaschewski** [comment]
If I understand correctly, you have a list of structures like that:
```json
[
  {
    "name" : "Peter",
    "age" : 23
  },
  {
    "name" : "Mary",
    "age" : 21
  },
  {
    "name" : "Barky",
    "age" : 5
  }
]
```

And you want to calculate the sum of all the `age` fields?

Then a function that does this could go like the following implementation (not tested).
It gets two parameters, the list of structures and the attribute name that should be calculated:
```java
  @FunctionParameter(name = "list")
  @FunctionParameter(name = "attributeName")
  public class SumAllFunction extends AbstractFunction {
    @Override
    public EvaluationValue evaluate(Expression expression, Token functionToken, EvaluationValue... parameterValues) throws EvaluationException {
      List<EvaluationValue> list = parameterValues[0].getArrayValue();
      String attributeName = parameterValues[1].getStringValue();

      BigDecimal sum = BigDecimal.ZERO;
      for (EvaluationValue entry : list) {
        Map<String, EvaluationValue> map = entry.getStructureValue();
        if (map.containsKey(attributeName)) {
          sum = sum.add(map.get(attributeName).getNumberValue());
        }
      }
      return EvaluationValue.numberValue(sum);
    }
  }
```

Then doing a `SUMALL(array, "age")` should return 49.

It probably needs some type checks, like is it an array, is each element a structure and i the attribute a number.

**Comment by lg32810** [comment]
Thanks a lot. That works.
